src/libos/src/net/socket/uring/datagram/ip_packet.rs

```diff
@@ -0,0 +1,318 @@
+        Ok(Self {
+            common,
+            is_connected,
+            proto,
+            sender,
+            receiver,
+        })
+    }
+
+    pub fn domain(&self) -> Domain {
+        A::domain()
+    }
+
+    pub fn host_fd(&self) -> FileDesc {
+        self.common.host_fd()
+    }
+
+    pub fn status_flags(&self) -> StatusFlags {
+        // Only support O_NONBLOCK
+        if self.common.nonblocking() {
+            StatusFlags::O_NONBLOCK
+        } else {
+            StatusFlags::empty()
+        }
+    }
+
+    pub fn set_status_flags(&self, new_flags: StatusFlags) -> Result<()> {
+        // Only support O_NONBLOCK
+        let nonblocking = new_flags.is_nonblocking();
+        self.common.set_nonblocking(nonblocking);
+        Ok(())
+    }
+
+    pub fn bind(&self, addr: &A) -> Result<()> {
+        do_bind(self.host_fd(), addr)?;
+        self.common.set_addr(addr);
+
+        Ok(())
+    }
+
+    // ip packet socket don't support getpeername, and if we connect address is unspec,
+    // we mark it disconnected
+    pub fn connect(&self, peer_addr: Option<&A>) -> Result<()> {
+        if let Some(peer) = peer_addr {
+            do_connect(self.host_fd(), Some(peer))?;
+
+            self.receiver.reset_shutdown();
+            self.sender.reset_shutdown();
+            self.common.set_peer_addr(peer);
+
+            self.is_connected.store(true, Ordering::Release);
+        } else {
+            do_connect::<A>(self.host_fd(), None)?;
+            self.common.reset_peer_addr();
+            self.is_connected.store(false, Ordering::Release);
+        }
+        Ok(())
+    }
+
+    pub fn read(&self, buf: &mut [u8]) -> Result<usize> {
+        self.readv(&mut [buf])
+    }
+
+    pub fn readv(&self, bufs: &mut [&mut [u8]]) -> Result<usize> {
+        self.recvmsg(bufs, RecvFlags::empty(), None)
+            .map(|(ret, ..)| ret)
+    }
+
+    pub fn recvmsg(
+        &self,
+        bufs: &mut [&mut [u8]],
+        flags: RecvFlags,
+        control: Option<&mut [u8]>,
+    ) -> Result<(usize, Option<A>, MsgFlags, usize)> {
+        self.receiver.recvmsg(bufs, flags, control)
+    }
+
+    pub fn write(&self, buf: &[u8]) -> Result<usize> {
+        self.writev(&[buf])
+    }
+
+    pub fn writev(&self, bufs: &[&[u8]]) -> Result<usize> {
+        self.sendmsg(bufs, None, SendFlags::empty(), None)
+    }
+
+    pub fn sendmsg(
+        &self,
+        bufs: &[&[u8]],
+        addr: Option<&A>,
+        flags: SendFlags,
+        control: Option<&[u8]>,
+    ) -> Result<usize> {
+        if self.proto == ProtocolType::RAW {
+            // INET6+SOCK_RAW+IPPROTO_RAW sockets can be created, but not written to.
+            if A::domain() == Domain::INET6 {
+                return_errno!(EINVAL, "raw socket with IPPROTO_RAW cannot be written to")
+            }
+
+            // SOCK_RAW+IPPROTO_RAW sockets need to include ip header
+            let buf_len: usize = bufs.iter().map(|buf| buf.len()).sum();
```

Comment:
test？

Reply:
Tested：
INET6 + IPPROTO_RAW socket can be created and written, but the buffer need to more than 40(IPV6 header size) bytes